### ISSUE TITLE
Shake invokes my rule with exceptions masked!

### ISSUE THREAD

**Issue by ezyang** [ndmitchell/shake#474]
Steps to reproduce:
1. Download and build https://github.com/ezyang/ghc-shake/ following the instructions (in particular need a patched version of Shake)
2. Apply this patch to ghc-shake

```
diff --git a/GhcShake.hs b/GhcShake.hs
index 9ae1d23..1d97241 100644
--- a/GhcShake.hs
+++ b/GhcShake.hs
@@ -268,7 +268,8 @@ doShake args srcs = do
     -- name.

     buildModuleRule $ \bm@(BuildModule raw_file mod is_boot) -> do
-
+        x <- liftIO $ getMaskingState
+        liftIO $ print x
         -- Make sure we rebuild if -this-unit-id changes
         _ <- askThisPackage
```

and build.
1. Run `ghc-shake --make All.hs -O -fforce-recomp -j3 +RTS -N3` with the following modules:

```
-- A.hs
module A where
-- B.hs
module B where
-- All.hs
module All where
import A
import B
```

Expected result: masking state is reported to be unmasked in all cases.

Actual result: masking state is unmasked for one, masked for two others!

Sorry for not having a reduced test case, more investigation needed.

**Comment by ndmitchell** [comment]
Eeek!

I just modified Shake to raise an error if any rules are started with exceptions unmasked. The test suite passed fine. Therefore, one of:

* It's something you are doing in your system that I don't do in any of my tests.
* Or it's something I've fixed in HEAD, which has diverged somewhat since your version (HEAD is in a very fluid state, so I don't recommend you merge to it at this point). That said, I haven't really gone near exception stuff in the refactoring.

It's certainly something to fix, but if there's no obvious way to figure out what's going on your side, I suggest I finish refactoring HEAD, then you merge to that (which has other benefits anyway), then we see if it still occurs.

**Comment by ezyang** [comment]
OK, I was able to repro on my Shake branch using the introductory Shake example. Using this diagnostic patch:

```
diff --git a/src/Development/Shake/Pool.hs b/src/Development/Shake/Pool.hs
index ac758f5..bb3d237 100644
--- a/src/Development/Shake/Pool.hs
+++ b/src/Development/Shake/Pool.hs
@@ -98,6 +98,7 @@ emptyS n deterministic = S Set.empty n 0 0 $ newQueue deterministic
 step :: Pool -> (S -> NonDet S) -> IO ()
 step pool@(Pool var done) op = do
     let onVar act = modifyVar_ var $ maybe (return Nothing) act
+    (\s -> print ("z",s)) =<< getMaskingState
     onVar $ \s -> do
         s <- op s
         res <- maybe (return Nothing) (fmap Just) $ dequeue $ todo s
```

then I get this:

```
ezyang@sabre:~/Dev/ghc-shake$ /srv/code/ghc-shake/dist-newstyle/build/ghc-shake-0.1.0.0/build/shake-test/shake-test
("z",Unmasked)
("z",Unmasked)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
("z",MaskedInterruptible)
# gcc (for _build/run)
gcc: fatal error: no input files
compilation terminated.
("z",MaskedInterruptible)
("z",MaskedInterruptible)
Error when running Shake build system:
* _build/run
user error (Development.Shake.cmd, system command failed
Command: gcc -o _build/run
Exit code: 1
Stderr:
gcc: fatal error: no input files
compilation terminated.
)
```

Next step, I'll try with Shake master.

**Comment by ezyang** [comment]
And for Shake master, I see something that looks a lot better:

```
ezyang@sabre:~/Dev/shake-master$ /srv/code/shake-master/dist-newstyle/build/fooshake-0.1.0.0/build/fooshake/fooshake
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
Unmasked
# gcc (for _build/src/Test/Tup/hello.o)
src/Test/Tup/hello.c:2:20: fatal error: square.h: No such file or directory
compilation terminated.
MaskedInterruptible
Unmasked
Error when running Shake build system:
* _build/run
* _build/src/Test/Tup/hello.o
user error (Development.Shake.cmd, system command failed
Command: gcc -c src/Test/Tup/hello.c -o _build/src/Test/Tup/hello.o -MMD -MF _build/src/Test/Tup/hello.m
Exit code: 1
Stderr:
src/Test/Tup/hello.c:2:20: fatal error: square.h: No such file or directory
compilation terminated.
)
```

Though, there is one MaskedInterruptible snuck in there.

**Comment by ndmitchell** [comment]
Looking at the trace, you are tracing the step function of the thread pool, which is not where it runs user actions. I think you should be tracing:

```
diff --git a/src/Development/Shake/Internal/Core/Pool.hs b/src/Development/Shake/Internal/Core/Pool.hs
index 33ce16d..67c2137 100644
--- a/src/Development/Shake/Internal/Core/Pool.hs
+++ b/src/Development/Shake/Internal/Core/Pool.hs
@@ -116,7 +116,7 @@ step pool@(Pool var done) op = do
         case res of
             Just (now, todo2) | Set.size (threads s) < threadsLimit s -> do
                 -- spawn a new worker
-                t <- forkFinally (now >> worker pool) $ \res -> case res of
+                t <- forkFinally (getMaskingState >>= \s -> print s >> now >> worker pool) $ \res -> case res of
                     Left e -> onVar $ \s -> do
                         t <- myThreadId
                         mapM_ killThread $ Set.toList $ Set.delete t $ threads s
```

Since this is the place I actually run user actions, and there I only see Unmasked. In fact, because it always spawns a new thread, I can't see how you'd ever get a masked state for the users code.

Or have I got that wrong? Clearly something is masking somewhere...

**Comment by ndmitchell** [comment]
I finally found this and fixed it as 0a170059f4ffbdbc2678a866d4cf9d1eb1b5fe82. I was forking a thread, and forgot the mask status of the parent leaked into the masked state of the new thread. I can replicate this is a bigger program, but not in my test suite, so still working out how to test for it in general.

**Comment by ndmitchell** [comment]
I added a test, even though it doesn't fail normally. That will have to do...
